[SIL] [Parser] Move local value storage into its own class.

zoecarver opened the pull request:
Moves both the forwardReferences map and the localValues map into SILParsedLocalValueMap. This will allow it to be more portable so it can be used in other places later on.

zoecarver commented on include/swift/SIL/SILParser.h:
```diff
@@ -0,0 +1,67 @@
+
+class SILParsedLocalValueMap {
+  SILModule &module;
+  bool &hadError;
```
I really don't like that we have to have this here. I'd like if this class can be as close to a map as possible. It shouldn't worry about diagnostics and errors. Anyway, that's a change for later.

gottesmm replied:
Make this a callback so the interfaces are clean.

gottesmm replied:
Please make this a private header in ./lib/SIL/Parser. This doesn't need to be exposed outside the library.

zoecarver replied:
Good call. Made it a callback. I, unfortunately, had to expose the member and set it after initializing `SILParsedLocalValueMap` because you can't capture a member in a lambda in that class' constructor. 

What do you mean by a private header? Is that a CMake option?

gottesmm replied:
A private header is something in ./lib that isn't exposed. I would create a private .h file in ./lib/SIL/Parser. Maybe call it ./lib/SIL/Parser/SILParsedLocalValueMap.h

zoecarver replied:
Good to know. Will do.

zoecarver replied:
Done.

gottesmm reviewed (commented):
Please make these changes. I need to re-look at this afterwards.

gottesmm commented on lib/SIL/Parser/SILParsedLocalValueMap.h:
```diff
@@ -0,0 +1,71 @@
+  SILModule &module;
+
+  llvm::StringMap<SILValue> localValues;
+  llvm::StringMap<SourceLoc> forwardRefs;
```
Can you expand out forwardRefs -> forwardReferences and maybe add a comment on how it is used?

gottesmm replied:
Also, I noticed above that there was a note about non-determinism when iterating over forward refs. Why don't we just nip that in this PR as well by changing forwardRefs to be a MapVector.

zoecarver replied:
Good call. Done.

gottesmm commented on lib/SIL/Parser/SILParsedLocalValueMap.h:
```diff
@@ -0,0 +1,71 @@
+
+  /// Get a reference to the forward references stored in this class.
+  /// TODO: remove this method.
+  llvm::StringMap<SourceLoc> &getForwardRefs();
```
Then you can just return an array ref to the map vector here rather than leak out what the underlying type is.

gottesmm commented on lib/SIL/Parser/ParseSIL.cpp:
```diff
@@ -525,15 +595,15 @@ bool SILParser::diagnoseProblems() {
+  if (!localValueMap.getForwardRefs().empty()) {
     // FIXME: These are going to come out in nondeterministic order.
-    for (auto &Entry : ForwardRefLocalValues)
+    for (auto &Entry : localValueMap.getForwardRefs())
```
This is the comment I was talking about below.

zoecarver replied:
I'm planning to submit a patch to start removing forward refs from the test suite after this lands.

gottesmm reviewed (commented):
More comments.

gottesmm commented on lib/SIL/Parser/ParseSIL.cpp:
```diff
@@ -46,6 +48,79 @@ using namespace swift::syntax;
+                                               SILType type, SILLocation loc,
+                                               SILBuilder &builder) {
+  if (name.isUndef())
+    return SILUndef::get(type, builder.getFunction());
```
Why not just pass in the function instead of passing in the SILBuilder. In general passing around SILBuilder is an old style that we do not want to propagate unless we really need to.

zoecarver replied:
Good to know. Done.

zoecarver replied:
Actually, I remember now, when I originally wrote this patch I did try to make this a function. Unfortunately, the builder doesn't always have a function associated with it. We just hope that if we ever run into an `undef` or forward ref, there is a function (which is an error waiting to happen). I will fix this in another PR (because it requires a larger fix).

zoecarver committed:
[SIL] [Parser] Move local value storage into its own class.

Moves both the forwardReferences map and the localValues map into
SILParsedLocalValueMap.

CodaFi commented on lib/SIL/Parser/SILParsedLocalValueMap.h:
```diff
@@ -0,0 +1,72 @@
+                              Args &&... args);
+
+public:
+  llvm::function_ref<void()> hadError = []() {
```
function_ref members are deeply suspicious.  This should be a std::function or const (or both if you can swing it).

CodaFi commented on lib/SIL/Parser/SILParsedLocalValueMap.h:
```diff
@@ -0,0 +1,72 @@
+    llvm_unreachable("hadError must be initialized.");
+  };
+
+  SILParsedLocalValueMap(SILModule &module) : module(module) {}
```
`explicit` please

zoecarver commented:
@CodaFi thanks for the review. Fixed.

zoecarver committed:
Merge branch 'master' into sil-parser/local-val-storage

zoecarver commented:
Friendly ping. Any more comments?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

zoecarver commented:
@gottesmm ping.
